File: 0907-koko-eating-bananas/0907-koko-eating-bananas.cpp
class Solution {
public:
    long long totalhourstaken(vector<int>& nums, long long k) {
        long long totalhours = 0;
        for (long long bananas : nums) {
            totalhours += bananas / k;
            if (bananas % k) {
                totalhours += 1;
            }
        }
        return totalhours;
    }

    int findmax(vector<int>& nums) {
        long long largest = nums[0];
        for (int x : nums) {
            largest = max(largest, (long long)x);
        }
        return (int)largest;
    }

    int minEatingSpeed(vector<int>& piles, int h) {
        long long low = 1;
        long long high = findmax(piles);
        long long ans = high;
        while (low <= high) {
            long long mid = low + (high - low) / 2;
            long long hours = totalhourstaken(piles, mid);

            if (hours <= h) {
                ans = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return (int)ans;
    }
};

File: 0907-koko-eating-bananas/0907-koko-eating-bananas.java
class Solution {
    public boolean helper(int[] nums,int h, int k ){
        long totalhours = 0;
for (int num : nums) {
    if (num >= k) {
        totalhours += num / k;
    }
    if (num % k != 0) {
        totalhours += 1;
    }
}
return totalhours <= h ? true : false;

    }
    public int minEatingSpeed(int[] piles, int h) {
        int max = piles[0];
        for(int i =  0 ; i < piles.length ; i++){
           if(piles[i]>max) max = piles[i];
        }
        int start = 1;
        int end = max;
        int possible = end;
        while(start<=end){
            int mid = start + (end-start)/2;
            if(helper(piles,h,mid)){
                possible = mid;
                end = mid-1;
            }else{
                start = mid+1;
            }
        }
        return possible;
    }
}